### DISCUSSION TITLE
Teensy 4.1 and NewHaven EVE2. Touch working?

### DISCUSSION THREAD

**Discussion by rmills** [RudolphRiedel/FT800-FT813#41] [General]
**MCU:** Teensy 4.1
**Display:** NHD-5.0-800480FT-CTXL-T (EVE2)
**Display Link:** https://www.newhavendisplay.com/nhd50800480ftctxlt-p-9574.html
**Example Test:** FT800-FT813-5.x\examples\EVE_Test_Arduino_PlatformIO
**IDE:** PlatformIO

Demo needs some tweaks to work correctly, it seem like the config setup is correctly but the example won't work because the built in LED is shared with the SCK, pulling out the LED code made the example work. I don't have touch thou, i think the example has a button that should do something? Are more pins required? If so where are they defined? 

**Pins Used:**
EVE_CS 10     
EVE_PDN 9         
MOSI  11     
MISO 12     
SCK  13

main.cpp
```

#include <Arduino.h>
#include <SPI.h>

#include "EVE_target.h"
#include "EVE_commands.h"
#include "tft.h"

void setup()
{	
	Serial.begin(9600);
	while (!Serial) {
    ; // wait for serial port to connect. Needed for native USB
    }

 pinMode(EVE_CS, OUTPUT);
	digitalWrite(EVE_CS, HIGH);
	pinMode(EVE_PDN, OUTPUT);
	digitalWrite(EVE_PDN, LOW);

#if defined (ESP32) || (ARDUINO_NUCLEO_F446RE)
	EVE_init_spi();
#else
	SPI.begin(); /* sets up the SPI to run in Mode 0 and 1 MHz */
	SPI.beginTransaction(SPISettings(8000000, MSBFIRST, SPI_MODE0));
#endif

	TFT_init();

//SPI.beginTransaction(SPISettings(20000000, MSBFIRST, SPI_MODE0));

#if defined (ESP8266)
	SPI.setFrequency(16000000);
#endif
}

void loop()
{
	uint32_t current_millis;
	static uint32_t previous_millis = 0;
	static uint8_t display_delay = 0;
	uint32_t micros_start, micros_end;
	current_millis = millis();

	if((current_millis - previous_millis) > 4) /* execute the code every 5 milli-seconds */
	{
		previous_millis = current_millis;



		micros_start = micros();
		TFT_touch();
		micros_end = micros();
		num_profile_b = (micros_end - micros_start); /* calculate the micro-seconds passed during the call to TFT_touch */

		display_delay++;
		if(display_delay > 3) /* refresh the display every 20ms */
		{
			display_delay = 0;
			micros_start = micros();
			TFT_display();
			micros_end = micros();
			num_profile_a = (micros_end - micros_start); /* calculate the micro-seconds passed during the call to TFT_display */
		}
	}
}
```

platformio.ini
```
;PlatformIO Project Configuration File
;
;   Build options: build flags, source filter
;   Upload options: custom upload port, speed and extra flags
;   Library options: dependencies, extra library storages
;   Advanced options: extra scripting
;
; Please visit documentation for the other options and examples
; https://docs.platformio.org/page/projectconf.html

[env]
lib_deps = https://github.com/RudolphRiedel/FT800-FT813/archive/refs/heads/5.x.zip
framework = arduino
build_flags = -D EVE_NHD_50
check_tool = cppcheck
check_flags =
  cppcheck: --addon=cert.py --suppress=unusedFunction

[env:teensy41]
platform = teensy
board = teensy41
build_flags =
        ${env.build_flags}
         -D EVE_CS=10
         -D EVE_PDN=9
```

**Comment by RudolphRrr** [comment]
It has been a while since I tested with the Teensy 4.1.
If I recall correctly I ended up with removing the resistor for the internal LED.
Also the code for the LED should not interfere with SCK since the pin can not be both GPIO and SPI at the same time, but I need to check that as well.

>I don't have touch thou, i think the example has a button that should do something?

Yes, the ugly star begins to rotate when you press the "Touch!" button.

>Are more pins required?

No, but you probably need to calibrate the touch sensor.
Activate the code at line #210ff of tft.c and add a set of REG_TOUCH_TRANSFORM_x values for your NHD50 on top of the touch_calibrate() function. So one block for EVE_NHD_50 like there already is one for EVE_NHD_50.

**Comment by rmills** [comment]
Thank you, I will give that a try. If there is anything you want to test with the display/mcu/library combo please let me know.

**Comment by RudolphRiedel** [comment]
I have mechanical/electrical issues right now getting a Teensy 4.1 to work with an EVE3-50G.
Looks like my adapter boards are failing or my jumper cables are, or both.
I identified a board that definately works though by using it with an ATSAME51 of my own design and the non-Arduino example worked.
Then I switched to an Arduino Uno and got it work as well.
But using the same PlatformIO project for the Teensy 4.1 now with commented out LED code does not work, or at least not realiably so, it did show an image a couple of times but right now the display stays dark and accoring to my logic analyzer the MISO line does not work correctly.

**Reply by rmills** [reply]
I can say with two different screens, the first having the older RA8875 chip and the newer NewHaven screen. On the Teensy 4.1 they were very touchy with the connections, I was fighting jumpers constantly. With the RA8875 screen I could see and decode the SPI bus but the screen just did nothing. Never got it to work. With the EVE2 I did not have a 20pin box header so I had to make one with a cheap amazon kit I already had. Wasted 2 hours because something in the cheap pins and cables was loose, I just happen to touch it and see the screen flash on. Once I got everything secured down it worked consistently for me. 

I also missed early on I needed to set the "EVE_NHD_50" in the platformio.ini config, not sure why but it only worked when it was set in the first [env] global options, the [env:teensy41] build options did not seem to override it. I might have missed something else to set the board type thou. I don't know PlatformIO very well.

**Comment by RudolphRiedel** [comment]
There appears to be an issue around PlatformIO and I can not yet put my finger on what causes this.
The Aduino demo downloads the library sources from Github as dependancy,
defined in the platformio.ini file: lib_deps = https://github.com/RudolphRiedel/FT800-FT813/archive/refs/heads/5.x.zip
This ends up in multiple copies, one for each board, for example here:
/.pio/adafruit_metro_m4/EmbeddedVideoEngine/

I am using a Metro-M4 now and while it displays an first image just fine it instantly crashes.
The onboard LED is not blinking, the "Bytes" show "0", the "Time1" shows 0µs and touch is not working.
When I disable DMA though in EVE_target.h it works fine including touch.

Re-activate DMA, re-compile, upload, crash.

When I however copy all the source files from /.pio/adafruit_metro_m4/EmbeddedVideoEngine/ to the main /src folder,
delete the /.pio folder, comment out the lib_deps line from platformio.ini and build it again,
then it works fine with DMA.

This may be a linker issue, I try to find out more.

**Comment by RudolphRiedel** [comment]
Ok, I have freshly crimped jumper cables now and repaired two adapter boards with verifying the functionality of every pin.
I tried both boards with the Metro M4 and the EVE3-50G and this worked just fine (apart from the PlatformIO DMA issue that I have no figured out so far).
Now I combined the Teensy 4.1 with the fresh jumpers cables, the known-to-be-good adapter board and the EVE3-50G - and it does not work, the screen is inittialized, the backlight is switchted on, thats it.
So I connected a logic analyzer and it looks just perfect. From init after reset over single transfer to read the touch for example down to the display refresh every 20ms - but I am just looking at a blank screen.
Not even the calibration screen works.

I do not get it, this should be working just fine.
I maybe killed my Teensy but this would be in a very specific way like killing the MISO without anything else.

**Comment by RudolphRiedel** [comment]
>I maybe killed my Teensy but this would be in a very specific way like killing the MISO without anything else.

At work it dawned on me that this is not true either, if the MISO pin would be blown it would not make it thru init.
What the heck?

**Comment by rmills** [comment]
I have had some fights with the linker in PlatformIO, it has some annoying issues. For example if a folder is too deep, it just ignores it. I noticed it pulled the library for every board defined, so I removed all the boards I don't need from the ini and created a new clean project. I'm on windows and sometimes windows locks files and it takes a few try's to flash. I don't like how PlatformIO handles flash errors, if the teensy fails to flash and your not watching the logs you won't see the error because the tab closes. 

I assume you cut the 5V to USB trace on the teensy? I know the datasheet says the teensy can take less than 5V on that pin but I found the teensy is not stable below 5V, it will run but just has all sorts of issues random issues. Could be worth asking Paul on his forums if he has any ideas.

**Comment by RudolphRrr** [comment]
No, I did not cut the 5V to USB trace, that is the connection between VUSB and VIN?

I am just using one of these boards https://github.com/RudolphRiedel/EVE_display-adapter/tree/master/L-D5019-01-05
with an extra PSU to power the display as this is just a temporary setup.
I only replaced the 10R resistors on the input side with 56R and removed the 22pF caps to make it place nice with the Teensy 4.
And of course the GNDs are connected, I have a Y-cable just for this.
I have a new version of this board on the way that is more suitable for connecting to breakout boards, including an extra set of pin headers to connect a logic-analyzer to.
I probably release this in two weeks after it arrived and proved to be working, the schematic mostly is the same though, minus the part for the speaker that never worked to my satisfaction.

Also yes, I noticed as well that PlatformIO makes multiple copies of the dependancies but that the project builds for several boards is mostly an oddity anyways. :-)
From comparing the .elf files for the Metro-M4 build it looks like the DMA Handler is empty when building the project like it is meant to be build, that would make it indeed a linker issue.
Well, the idea with making my library PlatformIO compatible was to remove all the duplicate code from the repository, so I am still interested to make it work this way.

**Reply by RudolphRiedel** [reply]
As a first step at least the Metro-M4 example is now up and running, I just put in a pull-request for Adafruits ZeroDMA library.
I am not sure what exactly the cause of the issue is but removing the "weak" attribute from the DMAC handler functions in ZeroDMA fixes it.
This still looks like a linker issue in PlatformIO to me but I have no idea how to find the root cause of this, I am only glad that I found another way to fix this.
And my hope really was to find a hint on the way for what currently is wrong with the Teensy build.

**Comment by RudolphRiedel** [comment]
Well, I just set it all up again, Teensy 4.1, my new easier to use breakout board and the RiTFT43 that is configured in the example.
And now it just works - what the electron?
I even fetched a fresh copy of the example and changed the display to an EVE3-50G - works like a charm, DMA and everything.
Including the LED blinking code that I am about to remove since it won't do anything with most Arduino boards since the builtin LED shares a pin with the SPI.

Ok, regarding the Teensy 4.1, I do have the resistor for the LED removed from this board.
But this already was the case when I tried this before.

This is my new board:
[L-D5019-08-01_EVE.PDF](https://github.com/RudolphRiedel/FT800-FT813/files/8514369/L-D5019-08-01_EVE.PDF)

I will release this as well.
